File: packages/python/tryaii/diagnose/intake.py
"""Lenient inventory intake (SPEC.md §2.0).

Accepts loosely-shaped agent-written inventories and normalizes them without
ever guessing: a malformed site is skipped with a reason; a malformed
optional field is simply absent. Nothing here raises except a top-level
shape that cannot be an inventory at all.
"""

from __future__ import annotations

import re
from typing import Any, Optional

DEFAULT_OUTPUT_TOKENS = 500

_INT_RE = re.compile(r"-?\d+")
_NUM_RE = re.compile(r"-?\d+(\.\d+)?")


def _coerce_int(value: Any, minimum: int) -> Optional[int]:
    """int >= minimum from an int, integral float, or integer string."""
    if isinstance(value, bool):
        return None
    if isinstance(value, int):
        n = value
    elif isinstance(value, float) and value.is_integer():
        n = int(value)
    elif isinstance(value, str) and _INT_RE.fullmatch(value.strip()):
        n = int(value.strip())
    else:
        return None
    return n if n >= minimum else None


def _coerce_positive_number(value: Any) -> Optional[float]:
    """Finite number > 0 from a number or numeric string (ints stay integral
    through the SPEC §1.3 normalizer at emission time)."""
    if isinstance(value, bool):
        return None
    if isinstance(value, (int, float)):
        n = float(value)
    elif isinstance(value, str) and _NUM_RE.fullmatch(value.strip()):
        n = float(value.strip())
    else:
        return None
    return n if n > 0 and n == n and n not in (float("inf"), float("-inf")) else None


def _clean_str(value: Any) -> Optional[str]:
    if isinstance(value, str) and value.strip():
        return value.strip()
    return None


def _clean_prompt(value: Any) -> Any:
    """A prompt is a string or a cachelint prompt object; wrong types are
    treated as absent (never guessed)."""
    if isinstance(value, str) and value:
        return value
    if isinstance(value, dict):
        return value
    return None


def _clean_classification(value: Any) -> Optional[dict]:
    """Validate the `_classification` injection seam; invalid → absent."""
    if not isinstance(value, dict):
        return None
    sims = value.get("benchmark_similarities")
    if not isinstance(sims, dict) or not sims:
        return None
    clean_sims: dict[str, float] = {}
    for name, score in sims.items():
        if not isinstance(name, str) or isinstance(score, bool) \
                or not isinstance(score, (int, float)):
            return None
        clean_sims[name] = float(score)
    out: dict[str, Any] = {"benchmark_similarities": clean_sims}
    for key in ("broad_category", "subcategory"):
        s = _clean_str(value.get(key))
        if s is not None:
            out[key] = s
    for key in ("confidence", "difficulty"):
        v = value.get(key)
        if isinstance(v, (int, float)) and not isinstance(v, bool):
            out[key] = float(v)
    return out


def normalize_inventory(data: Any) -> dict:
    """Normalize an inventory (SPEC.md §2.0).

    Returns {"defaults": {"calls_per_day", "output_tokens"},
             "sites": [...], "skipped": [{"index", "reason"}]}.
    Raises ValueError only when the top-level shape is not an inventory.
    """
    if isinstance(data, list):
        raw_sites, raw_defaults = data, {}
    elif isinstance(data, dict) and isinstance(data.get("sites"), list):
        raw_sites = data["sites"]
        raw_defaults = data.get("defaults") if isinstance(data.get("defaults"), dict) else {}
    else:
        raise ValueError(
            "inventory must be an object with a 'sites' array, or an array of sites")

    defaults = {
        "calls_per_day": _coerce_positive_number(raw_defaults.get("calls_per_day")),
        "output_tokens": (
            _coerce_int(raw_defaults.get("output_tokens"), 1) or DEFAULT_OUTPUT_TOKENS),
    }

    sites: list[dict] = []
    skipped: list[dict] = []
    seen_ids: dict[str, int] = {}

    for index, raw in enumerate(raw_sites):
        if not isinstance(raw, dict):
            skipped.append({"index": index, "reason": "site is not an object"})
            continue
        file = _clean_str(raw.get("file"))
        if file is None:
            skipped.append({"index": index, "reason": "missing file"})
            continue
        line = _coerce_int(raw.get("line"), 1)
        if line is None:
            skipped.append({"index": index, "reason": "missing or invalid line"})
            continue

        site_id = _clean_str(raw.get("id")) or f"{file}:{line}"
        count = seen_ids.get(site_id, 0) + 1
        seen_ids[site_id] = count
        if count > 1:
            site_id = f"{site_id}#{count}"

        sites.append({
            "site_id": site_id,
            "index": index,
            "file": file,
            "line": line,
            "prompt": _clean_prompt(raw.get("prompt")),
            "provider": _clean_str(raw.get("provider")),
            "model": _clean_str(raw.get("model")),
            "calls_per_day": _coerce_positive_number(raw.get("calls_per_day")),
            "output_tokens": _coerce_int(raw.get("output_tokens"), 1),
            "notes": _clean_str(raw.get("notes")),
            "classification": _clean_classification(raw.get("_classification")),
        })

    return {"defaults": defaults, "sites": sites, "skipped": skipped}

File: packages/node/src/diagnose/intake.ts
/**
 * Lenient inventory intake (shared/diagnose/SPEC.md §2.0).
 *
 * Accepts loosely-shaped agent-written inventories and normalizes them
 * without ever guessing: a malformed site is skipped with a reason; a
 * malformed optional field is simply absent. Mirrors the Python reference
 * (diagnose/intake.py) byte-for-byte in every emitted value.
 */

export const DEFAULT_OUTPUT_TOKENS = 500;

const INT_RE = /^-?\d+$/;
const NUM_RE = /^-?\d+(\.\d+)?$/;

export interface NormalizedSite {
  site_id: string;
  index: number;
  file: string;
  line: number;
  prompt: unknown;
  provider: string | null;
  model: string | null;
  calls_per_day: number | null;
  output_tokens: number | null;
  notes: string | null;
  classification: Record<string, unknown> | null;
}

export interface NormalizedInventory {
  defaults: { calls_per_day: number | null; output_tokens: number };
  sites: NormalizedSite[];
  skipped: { index: number; reason: string }[];
}

/** int >= minimum from an int, integral float, or integer string. */
function coerceInt(value: unknown, minimum: number): number | null {
  let n: number;
  if (typeof value === 'number' && Number.isFinite(value) && Number.isInteger(value)) {
    n = value;
  } else if (typeof value === 'string' && INT_RE.test(value.trim())) {
    n = parseInt(value.trim(), 10);
  } else {
    return null;
  }
  return n >= minimum ? n : null;
}

/** Finite number > 0 from a number or numeric string. */
function coercePositiveNumber(value: unknown): number | null {
  let n: number;
  if (typeof value === 'number') {
    n = value;
  } else if (typeof value === 'string' && NUM_RE.test(value.trim())) {
    n = parseFloat(value.trim());
  } else {
    return null;
  }
  return Number.isFinite(n) && n > 0 ? n : null;
}

function cleanStr(value: unknown): string | null {
  if (typeof value === 'string' && value.trim()) return value.trim();
  return null;
}

/** A prompt is a string or a cachelint prompt object; wrong types are absent. */
function cleanPrompt(value: unknown): unknown {
  if (typeof value === 'string' && value) return value;
  if (value !== null && typeof value === 'object' && !Array.isArray(value)) return value;
  return null;
}

/** Validate the `_classification` injection seam; invalid -> absent. */
export function cleanClassification(value: unknown): Record<string, unknown> | null {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) return null;
  const v = value as Record<string, unknown>;
  const sims = v.benchmark_similarities;
  if (
    sims === null ||
    typeof sims !== 'object' ||
    Array.isArray(sims) ||
    Object.keys(sims as object).length === 0
  ) {
    return null;
  }
  const cleanSims: Record<string, number> = {};
  for (const [name, score] of Object.entries(sims as Record<string, unknown>)) {
    if (typeof score !== 'number') return null;
    cleanSims[name] = score;
  }
  const out: Record<string, unknown> = { benchmark_similarities: cleanSims };
  for (const key of ['broad_category', 'subcategory'] as const) {
    const s = cleanStr(v[key]);
    if (s !== null) out[key] = s;
  }
  for (const key of ['confidence', 'difficulty'] as const) {
    const n = v[key];
    if (typeof n === 'number') out[key] = n;
  }
  return out;
}

/**
 * Normalize an inventory (SPEC.md §2.0). Throws only when the top-level
 * shape is not an inventory.
 */
export function normalizeInventory(data: unknown): NormalizedInventory {
  let rawSites: unknown[];
  let rawDefaults: Record<string, unknown>;
  if (Array.isArray(data)) {
    rawSites = data;
    rawDefaults = {};
  } else if (
    data !== null &&
    typeof data === 'object' &&
    Array.isArray((data as Record<string, unknown>).sites)
  ) {
    const d = data as Record<string, unknown>;
    rawSites = d.sites as unknown[];
    rawDefaults =
      d.defaults !== null && typeof d.defaults === 'object' && !Array.isArray(d.defaults)
        ? (d.defaults as Record<string, unknown>)
        : {};
  } else {
    throw new Error("inventory must be an object with a 'sites' array, or an array of sites");
  }

  const defaults = {
    calls_per_day: coercePositiveNumber(rawDefaults.calls_per_day),
    output_tokens: coerceInt(rawDefaults.output_tokens, 1) ?? DEFAULT_OUTPUT_TOKENS,
  };

  const sites: NormalizedSite[] = [];
  const skipped: { index: number; reason: string }[] = [];
  const seenIds = new Map<string, number>();

  for (let index = 0; index < rawSites.length; index++) {
    const raw = rawSites[index];
    if (raw === null || typeof raw !== 'object' || Array.isArray(raw)) {
      skipped.push({ index, reason: 'site is not an object' });
      continue;
    }
    const site = raw as Record<string, unknown>;
    const file = cleanStr(site.file);
    if (file === null) {
      skipped.push({ index, reason: 'missing file' });
      continue;
    }
    const line = coerceInt(site.line, 1);
    if (line === null) {
      skipped.push({ index, reason: 'missing or invalid line' });
      continue;
    }

    let siteId = cleanStr(site.id) ?? `${file}:${line}`;
    const count = (seenIds.get(siteId) ?? 0) + 1;
    seenIds.set(siteId, count);
    if (count > 1) siteId = `${siteId}#${count}`;

    sites.push({
      site_id: siteId,
      index,
      file,
      line,
      prompt: cleanPrompt(site.prompt),
      provider: cleanStr(site.provider),
      model: cleanStr(site.model),
      calls_per_day: coercePositiveNumber(site.calls_per_day),
      output_tokens: coerceInt(site.output_tokens, 1),
      notes: cleanStr(site.notes),
      classification: cleanClassification(site._classification),
    });
  }

  return { defaults, sites, skipped };
}
